packages/core/contracts/UnirepSocial.sol

```diff
@@ -306,4 +317,34 @@ contract UnirepSocial {
     ) external {
         unirep.updateUserStateRoot(publicSignals, proof, proofIndexRecords);
     }
+
+     /*
+     * Set new user name for an epochKey
+     * @param epochKey epoch key that attempts to set a new uername
+     * @param oldUsername
+     * @param newUsername requested new user name
+     */
+     function setUsername(
+        uint256 epochKey,
+        uint256 oldUsername,
+        uint256 newUsername
+     ) external {
+        // check if the new username is taken
+        require(isClaimed[newUsername] == false, "This user name is already taken");
+
+        count = userNames.length;
+        for(i=0; i<count;i++) {
+            if(userNmaes[i] == oldUsername) {
+                // free the old username
+                isClaimed[userNmaes[i]] = false;
```

Comment:
This internal logic is correct but we don't need the loop around it. You can modify `isClaimed` using `oldUsername` and `newUsername`. e.g.

```
isClaimed[oldUsername] = false;
isClaimed[newUsername] = true;
```

Generally we want to avoid iterating over arrays in contracts, especially when they can grow to large sizes. This function would eventually be impossible to execute because it would have so many iterations that it would exceed the block gas limit.